sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/throughputControl/controller/group/ThroughputGroupControllerBase.java

```diff
@@ -0,0 +1,238 @@
+        IThroughputRequestController requestController;
+        if (this.connectionMode == ConnectionMode.DIRECT) {
+            requestController = new PkRangesThroughputRequestController(
+                this.globalEndpointManager,
+                this.partitionKeyRangeCache,
+                this.targetContainerRid,
+                this.groupThroughput.get());
+
+        } else if (this.connectionMode == ConnectionMode.GATEWAY) {
+            requestController = new GlobalThroughputRequestController(this.globalEndpointManager, this.groupThroughput.get());
+        } else {
+            throw new IllegalArgumentException(String.format("Connection mode %s is not supported"));
+        }
+
+        return requestController.init();
+    }
+
+    public boolean isUseByDefault() {
+        return this.group.isUseByDefault();
+    }
+
+    public void onContainerMaxThroughputRefresh(int maxContainerThroughput) {
+        if (this.maxContainerThroughput.getAndSet(maxContainerThroughput) != maxContainerThroughput) {
+            this.calculateGroupThroughput();
+        }
+    }
+
+    @Override
+    public Mono<Void> close() {
+        this.cancellationTokenSource.cancel();
+        return this.resolveRequestController()
+            .flatMap(requestController -> requestController.close());
+    }
+
+    @Override
+    public <T> Mono<T> processRequest(RxDocumentServiceRequest request, Mono<T> nextRequestMono) {
+        return this.resolveRequestController()
+            .flatMap(requestController -> {
+                if (requestController.canHandleRequest(request)) {
+                    return Mono.just(requestController);
+                } else {
+                    // We can not find the matching pkRange, it is because either the group control is out of sync
+                    // or the request has staled info.
+                    // We will handle the first scenario by creating a new request controller,
+                    // while for second scenario, we will go to the original request mono, which will eventually get exception from server
+                    return this.shouldUpdateRequestController(request)
+                        .flatMap(shouldUpdate -> {
+                            if (shouldUpdate) {
+                                requestController.close();
+                                this.refreshRequestController();
+                                return this.resolveRequestController();
+                            } else {
+                                return Mono.just(requestController);
+                            }
+                        });
+                }
+            })
+            .flatMap(requestController -> {
+                if (requestController.canHandleRequest(request)) {
+                    return requestController.processRequest(request, nextRequestMono)
+                        .doOnError(throwable -> this.doOnError(throwable));
```

Comment:
I suggest we change the internal `doOnError()` name to something else, so that it doesn't conflicts with reactor -> `doOnError`

Reply:
rename to handleException